cmd/cmd_util.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright © 2021 The Persistent-Volume-Migrator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"strings"
+
+	"persistent-volume-migrator/internal/ceph/rbd"
+	"persistent-volume-migrator/internal/kubernetes"
+	logger "persistent-volume-migrator/internal/log"
+
+	v1 "k8s.io/api/core/v1"
+	k8s "k8s.io/client-go/kubernetes"
+)
+
+// migratePVC migrates a list of PVCs to CSI.
+func migratePVC(client *k8s.Clientset, pvcs *[]v1.PersistentVolumeClaim) error {
+	var err error
+	for _, pvc := range *pvcs {
+		pv, err := kubernetes.GetPV(client, pvc.Spec.VolumeName)
+		if err != nil {
+			return fmt.Errorf("failed to get PV object with name %s: %v", pvc.Spec.VolumeName, err)
+		}
+		logger.DefaultLog("PV found %v ", pv)
+		err = kubernetes.UpdateReclaimPolicy(client, pv)
+		if err != nil {
+			return fmt.Errorf("failed to update ReclaimPolicy for PV object %s: %v", pvc.Spec.VolumeName, err)
+		}
+		// Retrive old ceph volume name from PV
+		rbdImageName := kubernetes.GetFlexVolumeName(pv)
+		if rbdImageName == "" {
+			return fmt.Errorf("rbdImageName cannot be empty in Flex PV object: %v", pv)
+		}
+		logger.DefaultLog("flex rbdImageName name %v ", rbdImageName)
+
+		// Delete the PVC object
+		err = kubernetes.DeletePVC(client, &pvc) // nolint:gosec // skip gosec as pvc is accessed via it's reference.
+		if err != nil {
+			// TODO have an option to revert back if some Delete fails?
+			return fmt.Errorf("failed to Delete PVC object %s : %v", pvc.Name, err)
+		}
+		// Create new PVC with same name in destination storageclass
+		csiPVC := kubernetes.GenerateCSIPVC(destinationStorageClass, &pvc) // nolint:gosec // skip gosec as pvc is accessed via it's reference.
+		logger.DefaultLog("structure of the generated PVC in destination Storageclass \n %v ", csiPVC)
+
+		// Create new CSI PVC
+		csiPV, err := kubernetes.CreatePVC(client, csiPVC, 5)
+		if err != nil {
+			return fmt.Errorf("failed to Create CSI PVC object %s : %v", pvc.Name, err)
+		}
+		logger.DefaultLog("new PVC with same name created in CSI. Persistent Volume: %v ", csiPV)
+
+		// Extract new volume name from CSI PV
+		csiRBDImageName := kubernetes.GetCSIVolumeName(csiPV)
+		if csiRBDImageName == "" {
+			return fmt.Errorf("csiRBDImageName cannot be empty in PV object %v", csiPV)
```

Comment:
Why is this not returning an error?

Reply:
The function just returns the image name from the csi PV spec. One error we can return is in case the fetched value is empty?

Reply:
Yes, as an improvement for later, it will be more idiomatic Go for sure.